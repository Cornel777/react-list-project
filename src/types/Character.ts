export type Gender = 'male' | 'female' | 'n/a' | 'none';
export type Color = string;
export type URL = string;

export interface Character {
  name: string;
  height: string;
  mass: string;
  hair_color: Color;
  skin_color: Color;
  eye_color: Color;
  birth_year: string;
  gender: Gender;
  homeworld: URL;
  films: URL[];
  species: URL[];
  vehicles: URL[];
  starships: URL[];
  created: string;
  edited: string;
  url: URL;
}
