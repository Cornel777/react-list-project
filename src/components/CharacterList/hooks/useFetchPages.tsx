import { useQuery, keepPreviousData } from "@tanstack/react-query";
import fetchSwapi from "../../../services/api/fetchSwapi";
import { CharacterList } from "../../../types/CharacterList";

interface HookReturn {
  isFetching: boolean;
  isPlaceholderData: boolean;
  isPending: boolean;
  isError: boolean;
  error: Error | null;
  data: CharacterList | undefined;
}


const useFetchPages = ( page: number): HookReturn => {

  const { isFetching, isPlaceholderData, isPending, isError, error, data } = useQuery<CharacterList>({
    queryKey: ['characterList', page],
    queryFn: () => fetchSwapi(`people/?page=${page}`),
    staleTime: 5000,
    placeholderData: keepPreviousData,
  })

    return {
      isFetching,
      isPlaceholderData,
      isPending,
      isError,
      error,
      data,
    }
}

export default useFetchPages;