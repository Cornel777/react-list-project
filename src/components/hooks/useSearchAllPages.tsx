import { Character } from "../../types/Character";
import fetchSwapi from "../../services/api/fetchSwapi";
import { useQuery } from "@tanstack/react-query";
import { useState } from "react";
import debounce from 'lodash/debounce';

const PAGE_SIZE = 10;

interface HookReturn {
  resultCharacters: Character[] | undefined;
  isLoading: boolean;
  errorObject: Error | null;
  setSearchTerm: (name: string) => void;
}

const fetchAllCharacters = async (): Promise<Character[]> => {
  const firstPageForCount = await fetchSwapi('people/?page=1');
  const totalPages = Math.ceil(firstPageForCount.count / PAGE_SIZE);

  const requests = [];
  let page = 1;
  while ( page <= totalPages ) {
    requests.push(fetchSwapi(`people/?page=${page}`));
    page++;
  }

  const allResponses = await Promise.all(requests);
  return allResponses.flatMap((pageResponse) => pageResponse.results);
};

const useSearchAllPages = (searchTerm: string): HookReturn => {
  const [debouncedSearchTerm, setDebouncedSearchTerm] = useState(searchTerm);

  const debouncedSearch = debounce((searchTerm: string) => setDebouncedSearchTerm(searchTerm), 300);
    

  const {data: allCharactersList, isLoading, error} = useQuery<Character[], Error>({
    queryKey: ['allCharacters', 'all'],
    queryFn: fetchAllCharacters,
    staleTime: 10 * 60 * 1000, // 5 minutes
    gcTime: 60 * 60 * 1000, // 30 minutes
  });
  const filterredSearchResults =  (debouncedSearchTerm.trim() === '') 
  ? [] 
  : allCharactersList?.filter((character) => {
    return character.name.toLowerCase().includes(debouncedSearchTerm.toLowerCase()) 
  });
  
  console.log('HOOOK', filterredSearchResults);

  return {
    resultCharacters: filterredSearchResults,
    isLoading,
    errorObject: error,
    setSearchTerm: debouncedSearch,
  }
  };

export default useSearchAllPages;