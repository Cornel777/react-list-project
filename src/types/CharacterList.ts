import { Character } from "./Character";
import { URL } from "./Character";

export interface CharacterList {
  count: number;
  next: URL | null;
  previous: URL | null;
  results: Character[];
}