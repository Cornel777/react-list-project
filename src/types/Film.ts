import { URL } from './Character';

export interface Film {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  characters: URL[];
  planets: URL[];
  starships: URL[];
  vehicles: URL[];
  species: URL[];
  created: Date;
  edited: Date;
  url: URL;
}