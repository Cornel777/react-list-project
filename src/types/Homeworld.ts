import { URL } from "./Character";

export interface Homeworld {
  name: string;
  rotation_period: string;
  orbital_period: string;
  diameter: string;
  climate: string;
  gravity: string;
  terrain: string;
  surface_water: string;
  population: string;
  residents: string[];
  films: URL[];
  created: string;
  edited: string;
  url: URL;
}

