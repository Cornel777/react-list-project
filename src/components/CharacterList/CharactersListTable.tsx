import { TableContainer, Paper, Table, TableHead, TableBody } from "@mui/material";
import { FunctionComponent } from "react";
import { Character } from "../../types/Character";
import Homeworld from "../CharacterDetails/components/Homeworld/Homeworld";
import { StyledTableRow, StyledTableCell } from "./styled/styledListTable";
import getIdFromUrl from "./utils/getIdFromUrl";

interface CharactersListTableProps {
  characters: Character[];
  onCharacterNavigation: (id: string) => void;
  onKeyDownNavigation: (e: React.KeyboardEvent, id: string) => void;
}

const CharactersListTable: FunctionComponent<CharactersListTableProps> = ({ 
  characters,
  onCharacterNavigation,
  onKeyDownNavigation
}) => {

  return (
    <TableContainer component={Paper} style={{ maxWidth: '500px' }}>
          <Table size="medium" aria-label="Characters list table">
            <TableHead>
              <StyledTableRow>
                <StyledTableCell>No.</StyledTableCell>
                <StyledTableCell>Name</StyledTableCell>
                <StyledTableCell>Gender</StyledTableCell>
                <StyledTableCell>Home Planet</StyledTableCell>
              </StyledTableRow>
            </TableHead>
            <TableBody>
              {characters.map((character) => {
                const characterId = getIdFromUrl(character.url)
                  return (
                      <StyledTableRow 
                        key={characterId} 
                        tabIndex={0} 
                        onClick={() => onCharacterNavigation(characterId!)}
                        onKeyDown={(e) => onKeyDownNavigation(e, characterId!)}
                      >
                        <StyledTableCell>{characterId}</StyledTableCell>
                        <StyledTableCell>{character.name}</StyledTableCell>
                        <StyledTableCell>{character.gender}</StyledTableCell>
                        <StyledTableCell><Homeworld homeworldUrl={character.homeworld}/></StyledTableCell>
                      </StyledTableRow>
                  )
              })}
           </TableBody>
          </Table>
       </TableContainer>
  )
}

export default CharactersListTable;