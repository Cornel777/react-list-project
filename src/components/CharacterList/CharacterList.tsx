import { TextField, Typography } from '@mui/material'
import { useLocation, useNavigate } from 'react-router'
import LoadingDots from '../LoadingDots/LoadingDots'
import { FunctionComponent, useEffect, useState } from 'react'
import {type CharacterList } from '../../types/CharacterList'
import BreakElement from '../CharacterDetails/components/styled/BreakElement'
import useSearchAllPages from '../hooks/useSearchAllPages';
import Pagination from '../Pagination/Pagination'
import CharactersListTable from './CharactersListTable'
import useFetchPages from './hooks/useFetchPages'
import usePrefetchNextPage from './hooks/usePrefetchNextPage'

const CharacterList: FunctionComponent = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const [page, setPage] = useState(1);
  const [totalPageCount, setTotalPageCount] = useState(0);
  const [searchTerm, setSearchTerm] = useState('');
  
  const { data, isFetching, isPending, isError, error, isPlaceholderData } = useFetchPages(page);
  usePrefetchNextPage(data, isPlaceholderData, page);
  const { resultCharacters, setSearchTerm: setDebouncedSearch} = useSearchAllPages(searchTerm);

  useEffect(() => {
    if (data) {
      setTotalPageCount(Math.ceil(data.count / 10));
    }
  }, [data])


  
  useEffect(() => {
    if (location.state?.page) {
      setPage(location.state.page);
    }
  }, [location]);


  const handleListElementNavigation = ( characterId: string) => { 
    navigate(`/characters/${characterId}`, 
      { state: { page } }
    )
  };

  const handleKeyDown = (e: React.KeyboardEvent, characterId: string) => {
    if (e.key === 'Enter' || e.key === ' ') {
      handleListElementNavigation(characterId)
    }
  };
  
  if (isPending) {
    return <LoadingDots/>
 }
 if (isError) {
   return <p>`Failed to load data ${error && error.message}`</p>
 }
 if (!data) {
   return null;
 }

  return (
      <>
        <Typography variant="h3">Character List</Typography>
        <BreakElement/>
        <TextField 
          data-testid="character-list-search-input"
          id="outlined-basic"
          label="Search character name"
          variant="outlined" margin='normal' 
          fullWidth
          value={searchTerm}
          onChange={(e) => {
              setSearchTerm(e.target.value)
              setDebouncedSearch(e.target.value)
            } 
          }
        />
        { resultCharacters?.length ? 
          <CharactersListTable 
            characters={resultCharacters}
            onCharacterNavigation={handleListElementNavigation}
            onKeyDownNavigation={handleKeyDown}
        /> : 
        <>
          <CharactersListTable 
            characters={data.results}
            onCharacterNavigation={handleListElementNavigation}
            onKeyDownNavigation={handleKeyDown}
          />
          <BreakElement/>
          <Pagination
            currentPage={page}
            totalPages={totalPageCount}
            hasNextPage={data.next !== null}
            isLoading={isFetching}
            isPlaceholderData={isPlaceholderData}
            onPageChange={setPage}
          />
       </>
        }
      </>
  )
}

export default CharacterList;
