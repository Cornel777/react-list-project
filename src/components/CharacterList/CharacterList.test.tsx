import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import CharacterList from './CharacterList';
import useSearchAllPages from '../hooks/useSearchAllPages';
import { Character } from '../../types/Character';
import useFetchPages from './hooks/useFetchPages';
import { type CharacterList as Characters } from '../../types/CharacterList';

vi.mock('./hooks/useFetchPages');
vi.mock('../hooks/useSearchAllPages');
vi.mock('./hooks/usePrefetchNextPage');


const mockNavigate = vi.fn();
vi.mock('react-router',
   async () => {
    const actual = await vi.importActual('react-router');
    return {
      ...actual,
      useNavigate: () => mockNavigate,
      useLocation: vi.fn().mockReturnValue({ state: { page: 1 } }),
    }
  }
);  

const mockData = {
  count: 82,
  next: "https://swapi.dev/api/people/?page=2",
  previous: null,
  results: [
    { 
      name: 'Luke Skywalker',
      url: 'https://swapi.dev/api/people/1/',
      height: '172',
      gender: 'male',
      homeworld: 'url1'
    }
  ]
};

const mockSearchData = [
  { 
    name: 'Luke Skywalker search result',
    url: 'https://swapi.dev/api/people/1/',
    height: '172',
    gender: 'male',
    homeworld: 'url1'
  }
]


describe('CharacterList component', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      }
    }
  });

  function renderWithClient(ui: React.ReactElement) {
    const testQueryClient = queryClient;
    const { rerender, ...result } = render(
        <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    )
    return {
        ...result,
        rerender: (rerenderUi: React.ReactElement) =>
            rerender(
                <QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>
            ),
    }
  }

  // const renderComponent = () => {
  //  return render (
  //     <QueryClientProvider client={queryClient}>
  //       <BrowserRouter>
  //         <CharacterList />
  //       </BrowserRouter>
  //     </QueryClientProvider>
  //   )
  // }

  beforeEach(() => {
    queryClient.clear();
    mockNavigate.mockClear();
    vi.mocked(useFetchPages).mockReturnValue({
      data: mockData as Characters,
      isFetching: false,
      isPending: false,
      isError: false,
      error: null,
      isPlaceholderData: false
    });
    vi.mocked(useSearchAllPages).mockReturnValue({ 
      resultCharacters: [],
      setSearchTerm: vi.fn(),
      isLoading: false,
      errorObject: null });
  });

  it('should match the snapshot', () => {
    const result = renderWithClient(<CharacterList />);
    expect(result).toMatchSnapshot();
  });

  it('should render the character list', async () => {
    const result = renderWithClient(<CharacterList />);
    await waitFor(() => {
      expect(result.getByText('Luke Skywalker')).toBeInTheDocument();
    });
    expect(result.getByText('male')).toBeInTheDocument();
  });

  it('renders search input', () => {
    renderWithClient(<CharacterList />);
    expect(screen.getByLabelText('Search character name')).toBeInTheDocument();
  });

  it('navigates to character details with correct page state', async () => {
    renderWithClient(<CharacterList />);
    
    const characterRow = await screen.findByText('Luke Skywalker');
    fireEvent.click(characterRow);
    
    expect(mockNavigate).toHaveBeenCalledWith(
      '/characters/1',
      { state: { page: 1 } }
    );
  });

  it('navigates on key press', async () => {
    renderWithClient(<CharacterList />);
    
    const characterRow = await screen.findByText('Luke Skywalker');
    fireEvent.keyDown(characterRow, { key: 'Enter' });
    
    expect(mockNavigate).toHaveBeenCalledWith(
      '/characters/1',
      { state: { page: 1 } }
    );
  });
 
  it('displays search results', async () => {
    vi.mocked(useSearchAllPages).mockReturnValue({
      resultCharacters: mockSearchData as Character[],
      setSearchTerm: vi.fn(),
      isLoading: false,
      errorObject: null
    });

    renderWithClient(<CharacterList />);
    const searchInput = screen.getByTestId('character-list-search-input');
    await userEvent.type(searchInput, 'Luke');

    await waitFor(() => {
      expect(screen.getByText('Luke Skywalker search result')).toBeInTheDocument();
    });
  });

  it('shows regular list when search is empty', async () => {
    const result = renderWithClient(<CharacterList />);
    const searchInput = screen.getByLabelText('Search character name');
    await userEvent.clear(searchInput);

    await waitFor(() => {
      expect(result.getByText(/Current Page:/i)).toBeInTheDocument();
    });
  });

  it('handles search input changes', async () => {
    const mockSetSearchTerm = vi.fn();
    vi.mocked(useSearchAllPages).mockReturnValue({
      resultCharacters: [],
      setSearchTerm: mockSetSearchTerm,
      isLoading: false,
      errorObject: null
    });

    renderWithClient(<CharacterList />);
    const searchInput = screen.getByLabelText('Search character name');
    await userEvent.type(searchInput, 'Luke');

    expect(mockSetSearchTerm).toHaveBeenCalledWith('Luke');
  });
});