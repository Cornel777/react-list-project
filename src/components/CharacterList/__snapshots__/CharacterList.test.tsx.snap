// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterList component > should match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3
        class="MuiTypography-root MuiTypography-h3 css-13msox4-MuiTypography-root"
      >
        Character List
      </h3>
      <br
        class="sc-gyycJP"
      />
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-16bevx5-MuiFormControl-root-MuiTextField-root"
        data-testid="character-list-search-input"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="outlined-basic"
          id="outlined-basic-label"
        >
          Search character name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="outlined-basic"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-81qg8w"
            >
              <span>
                Search character name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); max-width: 500px;"
      >
        <table
          aria-label="Characters list table"
          class="MuiTable-root css-1xwxv7r-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-r0kehe-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
                scope="col"
              >
                No.
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
                scope="col"
              >
                Name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
                scope="col"
              >
                Gender
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
                scope="col"
              >
                Home Planet
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root css-r0kehe-MuiTableRow-root"
              tabindex="0"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
              >
                1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
              >
                Luke Skywalker
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
              >
                male
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
              >
                <div>
                  Loading...
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <br
        class="sc-gyycJP"
      />
      <div
        class="sc-eaUbBy"
      >
        Current Page: 
        1
        /
        9
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1l0eyhc-MuiButtonBase-root-MuiButton-root"
        disabled=""
        id=":r1:"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ArrowBackIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
         Previous Page
      </button>
       
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1l0eyhc-MuiButtonBase-root-MuiButton-root"
        id=":r2:"
        tabindex="0"
        type="button"
      >
        Next Page 
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ArrowForwardIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <h3
      class="MuiTypography-root MuiTypography-h3 css-13msox4-MuiTypography-root"
    >
      Character List
    </h3>
    <br
      class="sc-gyycJP"
    />
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-16bevx5-MuiFormControl-root-MuiTextField-root"
      data-testid="character-list-search-input"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        Search character name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
          id="outlined-basic"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-81qg8w"
          >
            <span>
              Search character name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); max-width: 500px;"
    >
      <table
        aria-label="Characters list table"
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-r0kehe-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              No.
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Gender
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
              scope="col"
            >
              Home Planet
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-r0kehe-MuiTableRow-root"
            tabindex="0"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              Luke Skywalker
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              male
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
            >
              <div>
                Loading...
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <br
      class="sc-gyycJP"
    />
    <div
      class="sc-eaUbBy"
    >
      Current Page: 
      1
      /
      9
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1l0eyhc-MuiButtonBase-root-MuiButton-root"
      disabled=""
      id=":r1:"
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="ArrowBackIosIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
        />
      </svg>
       Previous Page
    </button>
     
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1l0eyhc-MuiButtonBase-root-MuiButton-root"
      id=":r2:"
      tabindex="0"
      type="button"
    >
      Next Page 
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="ArrowForwardIosIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
